### PR TITLE
changes to allow non-primary unique key associations in HR

### PR DESCRIPTION
**Pull request by gavinking**
see https://github.com/hibernate/hibernate-reactive/issues/565

This is a draft, which I need to think more about. But here so at least @Sanne and @DavideD can see what I'm thinking about doing.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/loader/entity/AbstractEntityLoader.java [hunk 1]
```diff
@@ -61,6 +61,11 @@ public Object load(Serializable id, Object optionalObject, SharedSessionContract
 		return load( session, id, optionalObject, id, lockOptions, readOnly );
 	}
 
+	@Override
+	public Object load(Object id, SharedSessionContractImplementor session, LockOptions lockOptions) {
+		return load( session, id, null, null, lockOptions, null );
+	}
+
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/loader/entity/AbstractEntityLoader.java:68]: This is a replacement for `loadByUniqueKey()`.

File: hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java [hunk 2]
```diff
@@ -140,14 +139,6 @@ public EntityLoader(
...
 	}
 
-	public Object loadByUniqueKey(SharedSessionContractImplementor session, Object key) {
-		return loadByUniqueKey( session, key, null );
-	}
-
-	public Object loadByUniqueKey(SharedSessionContractImplementor session, Object key, Boolean readOnly) {
-		return load( session, key, null, null, LockOptions.NONE, readOnly );
-	}
-
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java:150]: These really weren't doing a lot of work, and HR can't override them.

File: hibernate-core/src/main/java/org/hibernate/loader/entity/UniqueEntityLoader.java [hunk 3]
```diff
@@ -60,4 +69,11 @@ default Object load(
 			Boolean readOnly) {
 		return load( id, optionalObject, session, lockOptions );
 	}
+
+	default Object load(
+			Object id,
+			SharedSessionContractImplementor session,
+			LockOptions lockOptions) {
+		throw new UnsupportedOperationException();
+	}
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/loader/entity/UniqueEntityLoader.java:78]: This is a replacement for `loadByUniqueKey()`. Unimplemented by default. But in future we could extend unique key loading to other `UniqueEntityLoader`s.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java [hunk 4]
```diff
@@ -255,7 +255,7 @@
 
 	private final EntityLoaderLazyCollection loaders = new EntityLoaderLazyCollection();
 
-	private volatile Map<String,EntityLoader> uniqueKeyLoaders;
+	private volatile Map<String,UniqueEntityLoader> uniqueKeyLoaders;
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:258]: This change is necessary to allow HR to provide its own reactive loaders.

**Comment by DavideD** [comment]
Shouldn't this be a GitHub draft?

**Comment by gavinking** [comment]
Yes I meant to do that but I forgot.

**Comment by gavinking** [comment]
So here's the thing: the overloaded `load()` operations of `UniqueEntityLoader` are really a bit of a mess at this point. So, should I take this as an opportunity to clean it up, or do I just add that one new `default` method and leave everything else alone? (And clean it up for H6.)

WDYT, @Sanne, @sebersole?

**Comment by Sanne** [comment]
> So here's the thing: the overloaded load() operations of UniqueEntityLoader are really a bit of a mess at this point. So, should I take this as an opportunity to clean it up, or do I just add that one new default method and leave everything else alone? (And clean it up for H6.)

I guess it depends what you plan?  We can of course trust you and have no problem with small refactorings, but I'd still prefer us being highly conservative with 5.x in case you have semantic changes in mind.

N.B. I'll be mostly offline this week as I'm moving home - but I'm sure you all know what you're doing

**Comment by sebersole** [comment]
This is already refactored in 6, so no worry about compatibility - its already something we planned on being not compatible.

However, these are public APIs and probably better to leave alone for 5.x unless there is a clear reason.

**Comment by gavinking** [comment]
OK, fine. So then are the changes I've proposed here OK?

I added a `default` method to `UniqueEntityLoader` and removed a couple of methods from `EntityLoader`. Does anyone object to those changes?

File: hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java (hunk 2 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java:150]: I'll change this to deprecate them, suggesting to use `load` instead. Ok?

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java:150]: Totally fine by me :)

**Comment by Sanne** [comment]
I'm a bit puzzled to see that @gavinking 's changes are conflicting with patches I wrote ~5 months ago. @gavinking may I suggest to pull changes sometimes :)

**Comment by gavinking** [comment]
Uh..... wha?

**Comment by gavinking** [comment]
According to GH this branch is only 16 commits behind master. (I sent it two weeks ago.)

**Comment by Sanne** [comment]
tracking as https://hibernate.atlassian.net/browse/HHH-14474

**Comment by Sanne** [comment]
> According to GH this branch is only 16 commits behind master. (I sent it two weeks ago.)

I got confused as I had made some changes recently as well - still there were conflicts, looks like with other improvements of yours that have been merged in the meantime. FWIW this will look a little different in branch 5.4.

**Comment by Sanne** [comment]
This was merged now.

**Comment by gavinking** [comment]
Thanks so much @Sanne, appreciated!!

**Comment by Sanne** [comment]
thanks @gavinking :)  Could you check your HR branch using a fresh build from ORM branch `5.4` ? 

I had to make some little changes here - should be equivalent to your original patch, but I'd rather verify this before we tag.

**Comment by gavinking** [comment]
On it.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java [hunk 5]
```diff
@@ -2469,17 +2469,17 @@ private void internalInitSubclassPropertyAliasesMap(String path, Iterator proper
...
 			String propertyName,
 			Object uniqueKey,
-			SharedSessionContractImplementor session) throws HibernateException {
-		return getAppropriateUniqueKeyLoader( propertyName, session ).loadByUniqueKey( session, uniqueKey );
+			SharedSessionContractImplementor session) {
+		return getAppropriateUniqueKeyLoader( propertyName, session ).load( uniqueKey, session, LockOptions.NONE );
 	}
 
-	private EntityLoader getAppropriateUniqueKeyLoader(String propertyName, SharedSessionContractImplementor session) {
+	protected UniqueEntityLoader getAppropriateUniqueKeyLoader(String propertyName, SharedSessionContractImplementor session) {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:2476]: @Sanne I think we might have lost this change:

`getAppropriateUniqueKeyLoader()` needs to be made visible.

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:2476]: Ok, I wondered about that. Isn't access to override `createUniqueKeyLoader` good enough for HR ?

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:2476]: Ok, I wondered about that. Isn't access to override `createUniqueKeyLoader` good enough for HR ?

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:2476]: No i need it.

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:2476]: Ok, done. Thanks!

**Pull request closed without merging**